File: Golang/Version2_Traductor/main.go
package main

import (
	"bytes"
	"encoding/json"
	"encoding/xml"
	"fmt"
	"io"
	"net/http"
	"net/url"
)

type Envelope struct {
	Body Body `xml:"Body"`
}

type Body struct {
	Response NumberToWordsResponse `xml:"NumberToWordsResponse"`
}

type NumberToWordsResponse struct {
	Result string `xml:"NumberToWordsResult"`
}

func convertirNumero(numero int) (string, error) {

	soap := fmt.Sprintf(`<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xmlns:xsd="http://www.w3.org/2001/XMLSchema"
xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">

<soap:Body>

<NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
<ubiNum>%d</ubiNum>
</NumberToWords>

</soap:Body>

</soap:Envelope>`, numero)

	req, err := http.NewRequest(
		"POST",
		"https://www.dataaccess.com/webservicesserver/NumberConversion.wso",
		bytes.NewBuffer([]byte(soap)),
	)

	if err != nil {
		return "", err
	}

	req.Header.Set("Content-Type", "text/xml; charset=utf-8")
	req.Header.Set(
		"SOAPAction",
		"http://www.dataaccess.com/webservicesserver/NumberToWords",
	)

	client := &http.Client{}

	resp, err := client.Do(req)

	if err != nil {
		return "", err
	}

	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var env Envelope

	xml.Unmarshal(body, &env)

	return env.Body.Response.Result, nil
}

func traducir(texto string) (string, error) {

	urlGoogle :=
		"https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q=" +
			url.QueryEscape(texto)

	resp, err := http.Get(urlGoogle)

	if err != nil {
		return "", err
	}

	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var datos interface{}

	err = json.Unmarshal(body, &datos)

	if err != nil {
		return "", err
	}

	arr := datos.([]interface{})
	a0 := arr[0].([]interface{})
	a1 := a0[0].([]interface{})

	return a1[0].(string), nil
}

func main() {

	var numero int

	fmt.Print("Ingrese un número: ")

	fmt.Scan(&numero)

	ingles, err := convertirNumero(numero)

	if err != nil {
		panic(err)
	}

	fmt.Println()
	fmt.Println("Número en inglés:")
	fmt.Println(ingles)

	espanol, err := traducir(ingles)

	if err != nil {
		panic(err)
	}

	fmt.Println()
	fmt.Println("Número en español:")
	fmt.Println(espanol)

}

File: Rust/Version2_Traductor/src/main.rs
use quick_xml::events::Event;
use quick_xml::Reader;
use reqwest::blocking::Client;
use serde_json::Value;
use std::io;
use urlencoding::encode;

fn main() {

    println!("Ingrese un número:");

    let mut numero = String::new();

    io::stdin().read_line(&mut numero).unwrap();

    let numero = numero.trim();

    let ingles = numero_a_ingles(numero);

    println!();
    println!("Número en inglés:");
    println!("{}", ingles);

    let espanol = traducir(&ingles);

    println!();
    println!("Número en español:");
    println!("{}", espanol);
}

fn numero_a_ingles(numero: &str) -> String {

    let soap = format!(
r#"<?xml version="1.0" encoding="utf-8"?>
<soap:Envelope
xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
xmlns:xsd="http://www.w3.org/2001/XMLSchema"
xmlns:soap="http://schemas.xmlsoap.org/soap/envelope/">

<soap:Body>

<NumberToWords xmlns="http://www.dataaccess.com/webservicesserver/">
<ubiNum>{}</ubiNum>
</NumberToWords>

</soap:Body>

</soap:Envelope>"#,
        numero
    );

    let client = Client::new();

    let respuesta = client
        .post("https://www.dataaccess.com/webservicesserver/NumberConversion.wso")
        .header(
            "SOAPAction",
            "\"http://www.dataaccess.com/webservicesserver/NumberToWords\"",
        )
        .header("Content-Type", "text/xml; charset=utf-8")
        .body(soap)
        .send()
        .unwrap()
        .text()
        .unwrap();

    let mut reader = Reader::from_str(&respuesta);

    reader.config_mut().trim_text(true);

    let mut resultado = String::new();

    loop {

        match reader.read_event() {

            Ok(Event::Start(ref e))
                if e.name().as_ref() == b"NumberToWordsResult"
                || e.name().as_ref() == b"m:NumberToWordsResult" =>
            {
                if let Ok(Event::Text(e)) = reader.read_event() {
                    resultado = e.decode().unwrap().to_string();
                    break;
                }
            }

            Ok(Event::Eof) => break,

            Err(_) => break,

            _ => {}
        }
    }

    resultado
}

fn traducir(texto: &str) -> String {

    let url = format!(
        "https://translate.googleapis.com/translate_a/single?client=gtx&sl=en&tl=es&dt=t&q={}",
        encode(texto)
    );

    let respuesta = Client::new()
        .get(url)
        .send()
        .unwrap()
        .text()
        .unwrap();

    let json: Value = serde_json::from_str(&respuesta).unwrap();

    json[0][0][0]
        .as_str()
        .unwrap_or("")
        .to_string()
}
